File: java edition/largestRectangleArea.java
import java.util.*;

public class largestRectangleArea{
    public static void main(String[] args){
        int[] test = new int[]{2, 1, 5, 6, 2, 3};
        System.out.println(largestRectangleArea(test));
    }

    public static int largestRectangleArea(int[] height){
        if(height == null || height.length == 0){
            return 0;
        }

        Stack<Integer> stack = new Stack<Integer>();

        int maxArea = 0;

        for(int i = 0 ; i <= height.length; i++){
            int curt = (i == height.length)? -1 : height[i];
            while(!stack.isEmpty() && curt <= height[stack.peek()] ){
                int h = height[stack.pop()];
                int w = stack.isEmpty()? i : (i - stack.peek() - 1);
                maxArea = Math.max(maxArea, h * w);
            }
            stack.push(i);
        }
        return maxArea;
    }
}

File: largestRectangleArea.py
# time analysis O(n)
# space analysis O(n)

class Solution:
    def largestRectangleArea(self,height):
        if not height:
            return 0
        stk = []
        maxArea = 0
        height.append(-1)

        for i in range(len(height)):
            if len(stk) == 0 or i == 0 or height[i] > height[i - 1]:
                stk.append((height[i] , i))
            else:
                while len(stk) != 0 and stk[-1][0] >= height[i]:
                    (h,index) = stk.pop()
                    print (h,index)
                    print maxArea
                    if len(stk) == 0:
                        width = index + 1
                    else:
                        width = i - stk[-1][1] -1
                    print (width,h)
                    maxArea = max(width * h , maxArea)
                stk.append((height[i], i))
        return maxArea



a = Solution()
test = [5, 4, 1, 2]
test1 = [2, 1, 5, 6, 2, 3]
test2 = [1, 1]
#  print a.largestRectangleArea(test)
print a.largestRectangleArea(test2)
